{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Two SNP regression analysis\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem\n",
    "We are interested in evaluating change in effect size estimate from\n",
    "1 SNP vs 2 SNP model. Specifically for given gene we want to take 2 SNPs:\n",
    "one with smallest p-value in brain tissues and the other with smallest p-value in non-brain tissues. \n",
    "Then for each tissue in GTEx we perform the following regression analyses (with covariates adjusted for):\n",
    "\n",
    "* Expression level vs. (SNP A + SNP B)\n",
    "* Expression level vs. SNP A\n",
    "* Expression level vs. SNP B\n",
    "\n",
    "We plan to plot for selected analysis the estimated effect sizes and standard error to\n",
    "show how effect estimates may change.\n",
    "\n",
    "@surbut has selected from GTEx analysis a list of 27 genes \n",
    "whose eQTL across brain tissues are in different directions\n",
    "compared to their effects in other tissues. (FIXME: \n",
    "link to `diffplot.html` shows why these genes are selected).\n",
    "\n",
    "## Procedures\n",
    "Analysis is coded in pipeline `2SNP.sos`. The `extract` workflow in the pipeline\n",
    "selected the following SNPs for these genes:\n",
    "```\n",
    "\"\",\"gene\",\"SNP_Brain\",\"SNP_Other\"\n",
    "\"1\",\"ENSG00000025772.7\",\"20_42804635_T_C_b37\",\"20_43589041_G_A_b37\"\n",
    "\"2\",\"ENSG00000056661.9\",\"17_36904739_GC_G_b37\",\"17_36904739_GC_G_b37\"\n",
    "\"3\",\"ENSG00000064012.17\",\"2_202184331_C_T_b37\",\"2_202145612_C_T_b37\"\n",
    "\"4\",\"ENSG00000089486.12\",\"16_4525265_C_T_b37\",\"16_4594671_C_T_b37\"\n",
    "\"5\",\"ENSG00000104472.5\",\"8_142442538_A_G_b37\",\"8_141526038_T_C_b37\"\n",
    "\"6\",\"ENSG00000108384.10\",\"17_57068108_T_C_b37\",\"17_56804230_G_A_b37\"\n",
    "\"7\",\"ENSG00000112977.11\",\"5_10761009_C_G_b37\",\"5_10680547_T_C_b37\"\n",
    "\"8\",\"ENSG00000120088.10\",\"17_44336651_T_C_b37\",\"17_44233589_T_C_b37\"\n",
    "\"9\",\"ENSG00000135744.7\",\"1_230849872_C_T_b37\",\"1_230849886_T_G_b37\"\n",
    "\"10\",\"ENSG00000136059.10\",\"3_38030899_G_A_b37\",\"3_38001432_C_T_b37\"\n",
    "\"11\",\"ENSG00000140265.8\",\"15_44277216_C_T_b37\",\"15_43671355_A_G_b37\"\n",
    "\"12\",\"ENSG00000145214.9\",\"4_967331_C_A_b37\",\"4_970724_A_T_b37\"\n",
    "\"13\",\"ENSG00000149054.10\",\"11_7725092_C_T_b37\",\"11_6962971_C_T_b37\"\n",
    "\"14\",\"ENSG00000160766.10\",\"1_155194980_T_C_b37\",\"1_155172379_G_A_b37\"\n",
    "\"15\",\"ENSG00000164124.6\",\"4_159150467_G_A_b37\",\"4_159092633_A_G_b37\"\n",
    "\"16\",\"ENSG00000177084.12\",\"12_133259752_C_T_b37\",\"12_133234741_C_CGTGT_b37\"\n",
    "\"17\",\"ENSG00000181240.9\",\"19_6436305_C_T_b37\",\"19_6436305_C_T_b37\"\n",
    "\"18\",\"ENSG00000187824.4\",\"17_10633358_C_T_b37\",\"17_10633181_A_G_b37\"\n",
    "\"19\",\"ENSG00000188732.6\",\"7_23719779_G_A_b37\",\"7_23733072_A_G_b37\"\n",
    "\"20\",\"ENSG00000189171.9\",\"1_153600778_G_A_b37\",\"1_153604519_G_A_b37\"\n",
    "\"21\",\"ENSG00000189316.3\",\"7_64364866_T_C_b37\",\"7_64334166_T_C_b37\"\n",
    "\"22\",\"ENSG00000198794.7\",\"15_74513228_T_C_b37\",\"15_75292699_C_T_b37\"\n",
    "\"23\",\"ENSG00000225439.2\",\"2_74373342_A_T_b37\",\"2_74368578_C_T_b37\"\n",
    "\"24\",\"ENSG00000249846.2\",\"3_129843468_G_T_b37\",\"3_129841378_G_A_b37\"\n",
    "\"25\",\"ENSG00000264247.1\",\"18_72256641_G_T_b37\",\"18_72260234_T_C_b37\"\n",
    "\"26\",\"ENSG00000267508.1\",\"19_44901056_TAA_T_b37\",\"19_44902190_G_A_b37\"\n",
    "\"27\",\"ENSG00000272030.1\",\"1_153021592_C_A_b37\",\"1_153605909_G_A_b37\"\n",
    "```\n",
    "To view workflows in the pipeline:\n",
    "\n",
    "```\n",
    "sos run code/2SNP.ipynb -h\n",
    "```\n",
    "To run a workflow:\n",
    "\n",
    "```\n",
    "sos run code/2SNP.ipynb workflow_name\n",
    "```\n",
    "\n",
    "## Results\n",
    "Result is organized in [this table](http://stephenslab.github.io/gtexresults_matrixash/TwoSNP.html) \n",
    "with rows being genes and columns tissues.\n",
    "Each cell entry has two sub-columns and each column has two numbers. The two\n",
    "numbers on the left column are effect size estimates from single SNP analysis for each SNP. \n",
    "The numbers on the right column are estimates from joint analysis. Significance levels are\n",
    "marked by asterisks in R style. These numbers are colored red when estimates have opposite signs \n",
    "from single SNP analysis yet same sign from joint analysis, and blue vise versa.\n",
    "\n",
    "It also generates [Supplemental Figure: Linkage Disequilibirum impacts estimates](http://stephenslab.github.io/gtexresults_matrixash/TwoSNPPlot.pdf)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "[global]\n",
    "parameter: tissues_brain = [\"Brain_Anterior_cingulate_cortex_BA24\",\n",
    "           \"Brain_Caudate_basal_ganglia\",\n",
    "           \"Brain_Cerebellar_Hemisphere\",\n",
    "           \"Brain_Cerebellum\",\n",
    "           \"Brain_Cortex\",\n",
    "           \"Brain_Frontal_Cortex_BA9\",\n",
    "           \"Brain_Hippocampus\",\n",
    "           \"Brain_Hypothalamus\",\n",
    "           \"Brain_Nucleus_accumbens_basal_ganglia\",\n",
    "           \"Brain_Putamen_basal_ganglia\"]\n",
    "parameter: tissues_other = [\"Adipose_Subcutaneous\",\n",
    "           \"Adipose_Visceral_Omentum\",\n",
    "           \"Adrenal_Gland\",\n",
    "           \"Artery_Aorta\",\n",
    "           \"Artery_Coronary\",\n",
    "           \"Artery_Tibial\",\n",
    "           \"Breast_Mammary_Tissue\",\n",
    "           \"Cells_EBV-transformed_lymphocytes\",\n",
    "           \"Cells_Transformed_fibroblasts\",\n",
    "           \"Colon_Sigmoid\",\n",
    "           \"Colon_Transverse\",\n",
    "           \"Esophagus_Gastroesophageal_Junction\",\n",
    "           \"Esophagus_Mucosa\",\n",
    "           \"Esophagus_Muscularis\",\n",
    "           \"Heart_Atrial_Appendage\",\n",
    "           \"Heart_Left_Ventricle\",\n",
    "           \"Liver\",\n",
    "           \"Lung\",\n",
    "           \"Muscle_Skeletal\",\n",
    "           \"Nerve_Tibial\",\n",
    "           \"Ovary\",\n",
    "           \"Pancreas\",\n",
    "           \"Pituitary\",\n",
    "           \"Prostate\",\n",
    "           \"Skin_Not_Sun_Exposed_Suprapubic\",\n",
    "           \"Skin_Sun_Exposed_Lower_leg\",\n",
    "           \"Small_Intestine_Terminal_Ileum\",\n",
    "           \"Spleen\",\n",
    "           \"Stomach\",\n",
    "           \"Testis\",\n",
    "           \"Thyroid\",\n",
    "           \"Uterus\",\n",
    "           \"Vagina\",\n",
    "           \"Whole_Blood\"]\n",
    "parameter: colors_brain = ['#eeee00'] * 10\n",
    "parameter: colors_other = [\n",
    "           \"#ff6600\",\n",
    "           \"#ffaa00\",\n",
    "           \"#33dd33\",\n",
    "           \"#ff5555\",\n",
    "           \"#ffaa99\",\n",
    "           \"#ff0000\",\n",
    "           \"#33cccc\",\n",
    "           \"#cc66ff\",\n",
    "           \"#aaeeff\",\n",
    "           \"#eebb77\",\n",
    "           \"#cc9955\",\n",
    "           \"#8b7355\",\n",
    "           \"#552200\",\n",
    "           \"#bb9988\",\n",
    "           \"#9900ff\",\n",
    "           \"#660099\",\n",
    "           \"#22ffdd\",\n",
    "           \"#aabb66\",\n",
    "           \"#99ff00\",\n",
    "           \"#aaaaff\",\n",
    "           \"#ffd700\",\n",
    "           \"#ffaaff\",\n",
    "           \"#995522\",\n",
    "           \"#aaff99\",\n",
    "           \"#dddddd\",\n",
    "           \"#0000ff\",\n",
    "           \"#7777ff\",\n",
    "           \"#555522\",\n",
    "           \"#778855\",\n",
    "           \"#ffdd99\",\n",
    "           \"#aaaaaa\",\n",
    "           \"#006600\",\n",
    "           \"#ff66ff\",\n",
    "           \"#ff5599\",\n",
    "           \"#ff00bb\"\n",
    "]\n",
    "parameter: genes = [\"ENSG00000025772.7\",\n",
    "           \"ENSG00000056661.9\",\n",
    "           \"ENSG00000064012.17\",\n",
    "           \"ENSG00000089486.12\",\n",
    "           \"ENSG00000104472.5\",\n",
    "           \"ENSG00000108384.10\",\n",
    "           \"ENSG00000112977.11\",\n",
    "           \"ENSG00000120088.10\",\n",
    "           \"ENSG00000135744.7\",\n",
    "           \"ENSG00000136059.10\",\n",
    "           \"ENSG00000140265.8\",\n",
    "           \"ENSG00000145214.9\",\n",
    "           \"ENSG00000149054.10\",\n",
    "           \"ENSG00000160766.10\",\n",
    "           \"ENSG00000164124.6\",\n",
    "           \"ENSG00000177084.12\",\n",
    "           \"ENSG00000181240.9\",\n",
    "           \"ENSG00000187824.4\",\n",
    "           \"ENSG00000188732.6\",\n",
    "           \"ENSG00000189171.9\",\n",
    "           \"ENSG00000189316.3\",\n",
    "           \"ENSG00000198794.7\",\n",
    "           \"ENSG00000225439.2\",\n",
    "           \"ENSG00000249846.2\",\n",
    "           \"ENSG00000264247.1\",\n",
    "           \"ENSG00000267508.1\",\n",
    "           \"ENSG00000272030.1\"]\n",
    "parameter: metaplot_genes = [\"ENSG00000264247.1\", \"ENSG00000267508.1\", \"ENSG00000145214.9\"]\n",
    "parameter: eqtl_db = \"/project/mstephens/data/internal_supp/gtex-v6-sumstat-hdf5/MatrixEQTLSumStats.h5\"\n",
    "parameter: data_prefix = \"/project/mstephens/data/external_private_supp/gtex-v6/eqtl_data/GTEx_Analysis_2015-01-12_eQTLInputFiles\"\n",
    "parameter: work_dir = \".\"\n",
    "parameter: lib_src = [\"/project/mstephens/gtex/scripts/SumstatQuery.R\"]\n",
    "tissues = tissues_brain + tissues_other\n",
    "gtex_colors = colors_brain + colors_other\n",
    "\n",
    "[extract_1]\n",
    "# Identify SNPs\n",
    "# Input: one gene name\n",
    "# Output: text file of 3 columns: gene, SNP A and SNP B\n",
    "output: \"TwoSNPs.dat\" \n",
    "R:\n",
    "  for (i in c(${lib_src!r,})) {\n",
    "      source(i)\n",
    "  }\n",
    "  genes = c(${genes!r,})\n",
    "  output = data.frame(matrix(NA, length(genes), 3))\n",
    "  colnames(output) = c(\"gene\", \"SNP_Brain\", \"SNP_Other\")\n",
    "  j = 1\n",
    "  for (i in genes) {\n",
    "     dat = -log10(GetSS(i, ${eqtl_db!r})$\"p-value\") \n",
    "     brain.col = which(colnames(dat) %in% c(${tissues_brain!r,}))\n",
    "     snp.brain = rownames(dat)[matxMax(dat[, brain.col])[1]]\n",
    "     non.brain.col = which(colnames(dat) %in% c(${tissues_other!r,}))\n",
    "     snp.non.brain = rownames(dat)[matxMax(dat[, non.brain.col])[1]]\n",
    "     output[j,] = c(i, snp.brain, snp.non.brain)\n",
    "     j = j + 1\n",
    "  }\n",
    "  write.csv(output, ${output!r})\n",
    "\n",
    "[extract_casp8_1]\n",
    "# Run specifically hand-picked SNPs from CASP8\n",
    "output: \"TwoSNPsCASP8.dat\"\n",
    "R:\n",
    "  output = t(data.frame(c(\"ENSG00000064012.17\", \"2_202176024_T_C_b37\", \"2_202163483_G_A_b37\")))\n",
    "  colnames(output) = c(\"gene\", \"SNP_Brain\", \"SNP_Other\")\n",
    "  write.csv(output, ${output!r})\n",
    "\n",
    "[extract_2, extract_casp8_2]\n",
    "# Extract data\n",
    "# Input: one gene name and 2 SNPs\n",
    "# Output: one data-set for every tissue\n",
    "input: for_each = [\"tissues\"]\n",
    "task: concurrent = True\n",
    "R:\n",
    "  GetCovar <- function() {\n",
    "      cmd <- paste(\"gzip -dc\", paste(${data_prefix!r}, \"_covariates/\", paste0(Tissue, \"_Analysis.covariates.txt.gz\"), sep = ''), '>', Covfile)\n",
    "      ##print(cmd)\n",
    "      system(cmd)\n",
    "  }\n",
    "  GetGE <- function(gene) {\n",
    "      GESource <- paste(${data_prefix!r}, \"_geneLevelNormalizedExpressionMatrices/\", paste0(Tissue, \"_Analysis.expr.txt.gz\"), sep = '')\n",
    "      cmd1 <- paste(\"zcat \", GESource, \" | head -1 >\", GEfile)\n",
    "      cmd2 <- paste(\"zcat \", GESource, \" | awk '$1 == var {print; exit}' var=\\\"\", gene, \"\\\" >> \", GEfile, sep = '')\n",
    "      ##print(cmd1)\n",
    "      ##print(cmd2)\n",
    "      system(cmd1)\n",
    "      system(cmd2)\n",
    "  }\n",
    "  GetSNP <- function(snps) {\n",
    "      SNPSource <- paste(${data_prefix!r}, \"_snpMatrices/\", paste0(Tissue, \"_Analysis.snps.txt.gz\"), sep = '')\n",
    "      SNPPattern <- paste('\"', '^(', paste(snps, collapse = \"|\"), \")\\\\b\", '\"', sep = '')\n",
    "      cmd1 <- paste(\"zcat \", SNPSource, \" | head -1 >\", SNPfile)\n",
    "      cmd2 <- paste(\"zcat \", SNPSource, \" | grep -E \", SNPPattern, \" >> \", SNPfile, sep = '')\n",
    "      ##print(cmd1)\n",
    "      ##print(cmd2)\n",
    "      system(cmd1)\n",
    "      system(cmd2)\n",
    "  }\n",
    "  \n",
    "  dat = read.csv(${input!r}, stringsAsFactors = F)[,-1]\n",
    "  Tissue = ${_tissues!r}\n",
    "  Covfile = paste(${work_dir!r}, paste0(Tissue, '_Covariates.txt'), sep = '/')\n",
    "  GetCovar()\n",
    "  for (i in 1:nrow(dat)) {\n",
    "      print(c(Tissue, dat[i, 1]))\n",
    "      GEfile = paste(${work_dir!r}, paste0(Tissue, '_', dat[i, 1], '_GE.txt'), sep = '/')\n",
    "      SNPfile = paste(${work_dir!r}, paste0(Tissue, '_', dat[i, 1], '_SNP.txt'), sep = '/')\n",
    "      GetGE(dat[i, 1])\n",
    "      GetSNP(dat[i, c(2,3)])\n",
    "  }\n",
    "\n",
    "[analyze_1]\n",
    "# Run analysis\n",
    "# Input: tissue specific data-set\n",
    "# Output: an R object of 2 SNP model analysis\n",
    "input: for_each = [\"tissues\", \"genes\"]\n",
    "output: expand_pattern(\"{_tissues}_{_genes}.rds\")\n",
    "task: concurrent = True\n",
    "R:\n",
    "  covar = t(read.table(paste(${work_dir!r}, paste0(${_tissues!r}, '_Covariates.txt'), sep = '/'), header = T)[,-1])\n",
    "  genotype = t(read.table(paste(${work_dir!r}, paste0(${_tissues!r}, '_', ${_genes!r}, '_SNP.txt'), sep = '/'), header = T)[, -1])\n",
    "  if (ncol(genotype) == 1) {\n",
    "     saveRDS(list(joint = NA, single = NA), ${_output!r})\n",
    "  } else {\n",
    "    colnames(genotype) = c(\"SNP1\", \"SNP2\")\n",
    "    phenotype = t(read.table(paste(${work_dir!r}, paste0(${_tissues!r}, '_', ${_genes!r}, '_GE.txt'), sep = '/'), header = T)[,-1])\n",
    "    model.full = summary(lm(phenotype ~ genotype + covar))$coef[c(2,3),]\n",
    "    model.1 = summary(lm(phenotype ~ genotype[,1] + covar))$coef[2,]\n",
    "    model.2 = summary(lm(phenotype ~ genotype[,2] + covar))$coef[2,]\n",
    "    saveRDS(list(joint = model.full, single = rbind(model.1, model.2)), ${_output!r})\n",
    "  }\n",
    "  \n",
    "[tabulate_1]\n",
    "# Make table summary \n",
    "# Input: All analysis results\n",
    "# Output: A table of size gene by tissue\n",
    "#  in each cell there are univarate and joint anlaysis\n",
    "output: 'TwoSNPs.html'\n",
    "python:\n",
    "\n",
    "  def get_star(p):\n",
    "    if p >= 0.05:\n",
    "       return ''\n",
    "    elif p < 0.0001:\n",
    "       return '***'\n",
    "    elif p < 0.001:\n",
    "       return '**'\n",
    "    else:\n",
    "       return '*'\n",
    "\n",
    "  # load results\n",
    "  from dsc.utils import load_rds\n",
    "  res = {'Genes': [${genes!r,}]}\n",
    "  for i in [${tissues!r,}]:\n",
    "      res[i] = []\n",
    "      for ii in [${genes!r,}]:\n",
    "          dat = load_rds('{}_{}.rds'.format(i, ii))\n",
    "          joint = '{0:.2f}{1}<br>{2:.2f}{3}'.format(dat['joint'][0, 0], get_star(dat['joint'][0, 3]),\n",
    "                dat['joint'][1, 0], get_star(dat['joint'][1, 3])) if len(dat['joint']) != 1 else 'NA'\n",
    "          single = '{0:.2f}{1}<br>{2:.2f}{3}'.format(dat['single'][0, 0], get_star(dat['single'][0, 3]),\n",
    "                dat['single'][1, 0], get_star(dat['single'][1, 3])) if len(dat['single']) != 1 else 'NA'\n",
    "          entry = '<table BORDER=\"3\" CELLPADDING=\"3\"><td>{}</td><td>{}</td></table>'.format(single, joint)\n",
    "          if joint != 'NA':\n",
    "             if dat['joint'][0, 0] * dat['joint'][1, 0] > 0 and dat['single'][0, 0] * dat['single'][1, 0] < 0:\n",
    "                entry = '<font color = \"red\">{}</font>'.format(entry)\n",
    "             if dat['joint'][0, 0] * dat['joint'][1, 0] < 0 and dat['single'][0, 0] * dat['single'][1, 0] > 0:\n",
    "                entry = '<font color = \"blue\">{}</font>'.format(entry)\n",
    "          res[i].append(entry)\n",
    "\n",
    "  # make table\n",
    "  keys = ['Genes'] + [${tissues!r,}]\n",
    "  length = len(res[keys[0]])\n",
    "  table = ['<table border=\"2\">', '<tr>']\n",
    "  for k in keys:\n",
    "    table.append('<td><b>%s</b></td>' % k)\n",
    "  table.append('</tr>')\n",
    "\n",
    "  for i in range(length):\n",
    "    table.append('<tr>')\n",
    "    for k in keys:\n",
    "      if k == 'Genes':\n",
    "        table.append('<td><i>%s</i></td>' % res[k][i])\n",
    "      else:\n",
    "        table.append('<td>%s</td>' % res[k][i])\n",
    "    table.append('</tr>')\n",
    "  table.append('</table>')\n",
    "  with open(${output!r}, 'w') as f:\n",
    "    f.write('<!DOCTYPE html><html><head><style>\\ntable {\\n    border-collapse: collapse;\\n    width: 100%;\\n}\\nth, td {\\n    text-align: center;\\n    padding: 8px;\\nfont-family: Sans-serif;\\n}\\ntr:nth-child(even){background-color: #f2f2f2}\\n</style></head><body><h2>Left in cell: single SNP analysis; right in cell: joint analysis</h2>')\n",
    "    f.write('\\n'.join(table) + '</body></html>')\n",
    "\n",
    "[metaplot_1]\n",
    "depends: R_library('rmeta')\n",
    "output: 'TwoSNPsMeta.pdf'\n",
    "R:\n",
    "  library(rmeta)\n",
    "  pdf(${output!r})\n",
    "  genes = c(${metaplot_genes!r,})\n",
    "  tissues = c(${tissues!r,})\n",
    "  gtex.colors = as.character(c(${gtex_colors!r,}))\n",
    "  for (i in genes) {\n",
    "      # beta\n",
    "      single_1b = vector()\n",
    "      single_2b = vector()\n",
    "      joint_1b = vector()\n",
    "      joint_2b = vector()\n",
    "      # SE\n",
    "      single_1e = vector()\n",
    "      single_2e = vector()\n",
    "      joint_1e = vector()\n",
    "      joint_2e = vector()\n",
    "      for (j in 1:length(tissues)) {\n",
    "          dat = readRDS(paste0(tissues[j], '_', i, '.rds'))\n",
    "          single_1b[j] = ifelse(!is.na(dat$single), dat$single[1,1], dat$single)\n",
    "          single_2b[j] = ifelse(!is.na(dat$single), dat$single[2,1], dat$single)\n",
    "          joint_1b[j] = ifelse(!is.na(dat$joint), dat$joint[1,1], dat$joint)\n",
    "          joint_2b[j] = ifelse(!is.na(dat$joint), dat$joint[2,1], dat$joint)\n",
    "          single_1e[j] = ifelse(!is.na(dat$single), dat$single[1,2], dat$single)\n",
    "          single_2e[j] = ifelse(!is.na(dat$single), dat$single[2,2], dat$single)\n",
    "          joint_1e[j] = ifelse(!is.na(dat$joint), dat$joint[1,2], dat$joint)\n",
    "          joint_2e[j] = ifelse(!is.na(dat$joint), dat$joint[2,2], dat$joint)\n",
    "      }\n",
    "      par(mfrow=c(2,2)) \n",
    "      metaplot(as.numeric(single_1b),as.numeric(single_1e),xlab = \"\",ylab=\"\",colors=meta.colors(box=gtex.colors), xlim=c(-1.5,1.5), main = paste0(i, \"\\nSNP 1 only\"))\n",
    "      metaplot(as.numeric(single_2b),as.numeric(single_2e),xlab = \"\",ylab=\"\",colors=meta.colors(box=gtex.colors), xlim=c(-1.5,1.5), main = paste0(i, \"\\nSNP 2 only\"))\n",
    "      metaplot(as.numeric(joint_1b),as.numeric(joint_1e),xlab = \"\",ylab=\"\",colors=meta.colors(box=gtex.colors), xlim=c(-1.5,1.5), main = paste0(i, \"\\nSNP 1 in joint analysis\"))\n",
    "      metaplot(as.numeric(joint_2b),as.numeric(joint_2e),xlab = \"\",ylab=\"\",colors=meta.colors(box=gtex.colors), xlim=c(-1.5,1.5), main = paste0(i, \"\\nSNP 2 in joint analysis\"))\n",
    "  }\n",
    "  dev.off()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos.jupyter.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  },
  "sos": {
   "default_kernel": "SoS",
   "kernels": [
    [
     "Python3",
     "python3",
     "Python3",
     "#FFE771"
    ],
    [
     "R",
     "ir",
     "R",
     "#DCDCDA"
    ],
    [
     "SoS",
     "sos",
     "",
     ""
    ]
   ],
   "panel": {
    "displayed": true,
    "height": 0,
    "style": "side"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
