<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Sarah Urbut" />


<title>Comparison of genes with tissue-specific eQTLs against other genes</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/readable.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">MASH analysis of GTEx</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Overview</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/stephenslab/gtexresults">source</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Comparison of genes with tissue-specific eQTLs against other genes</h1>
<h4 class="author"><em>Sarah Urbut</em></h4>

</div>


<p><strong>Last updated:</strong> 2018-05-24</p>
<strong>workflowr checks:</strong> <small>(Click a bullet for more information)</small>
<ul>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>R Markdown file:</strong> up-to-date </summary></p>
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Environment:</strong> empty </summary></p>
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Seed:</strong> <code>set.seed(1)</code> </summary></p>
<p>The command <code>set.seed(1)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Session information:</strong> recorded </summary></p>
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Repository version:</strong> <a href="https://github.com/stephenslab/gtexresults/tree/336a8b680ed49d0ab6e4373e82b87f2da6c919e4" target="_blank">336a8b6</a> </summary></p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
<pre><code>
Ignored files:
    Ignored:    .sos/
    Ignored:    data/.sos/
    Ignored:    gtex6_workflow_output/
    Ignored:    workflows/.ipynb_checkpoints/
    Ignored:    workflows/.sos/

</code></pre>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes. </details>
</li>
</ul>
<details> <summary> <small><strong>Expand here to see past versions:</strong></small> </summary>
<ul>
<table style="border-collapse:separate; border-spacing:5px;">
<thead>
<tr>
<th style="text-align:left;">
File
</th>
<th style="text-align:left;">
Version
</th>
<th style="text-align:left;">
Author
</th>
<th style="text-align:left;">
Date
</th>
<th style="text-align:left;">
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/stephenslab/gtexresults/blob/336a8b680ed49d0ab6e4373e82b87f2da6c919e4/analysis/Fig.ExpressionAnalysis.Rmd" target="_blank">336a8b6</a>
</td>
<td style="text-align:left;">
Peter Carbonetto
</td>
<td style="text-align:left;">
2018-05-24
</td>
<td style="text-align:left;">
wflow_publish(“Fig.ExpressionAnalysis.Rmd”)
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/stephenslab/gtexresults/blob/ae61358de3515c68c087a85738371adc8ed79a08/analysis/Fig.ExpressionAnalysis.Rmd" target="_blank">ae61358</a>
</td>
<td style="text-align:left;">
Peter Carbonetto
</td>
<td style="text-align:left;">
2018-05-24
</td>
<td style="text-align:left;">
wflow_publish(“Fig.ExpressionAnalysis.Rmd”)
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/stephenslab/gtexresults/blob/e8c6f6f601d66cc0c78c56bd892a3ecd869f6afc/analysis/Fig.ExpressionAnalysis.Rmd" target="_blank">e8c6f6f</a>
</td>
<td style="text-align:left;">
Peter Carbonetto
</td>
<td style="text-align:left;">
2018-05-24
</td>
<td style="text-align:left;">
Working on revisions to Fig.ExpressionAnalysis.Rmd.
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/stephenslab/gtexresults/blob/5ed67157e82a0368d5caa3d9a4bd45d1ddf75227/analysis/Fig.ExpressionAnalysis.Rmd" target="_blank">5ed6715</a>
</td>
<td style="text-align:left;">
Peter Carbonetto
</td>
<td style="text-align:left;">
2018-05-24
</td>
<td style="text-align:left;">
Moved AvgExpr..csv and working on changes to Fig.ExpressionAnalysis.Rmd analysis.
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/stephenslab/gtexresults/blob/80f285faf2eaf3c1e1b4fc25317c58996d5b2525/analysis/Fig.ExpressionAnalysis.Rmd" target="_blank">80f285f</a>
</td>
<td style="text-align:left;">
Gao Wang
</td>
<td style="text-align:left;">
2017-09-20
</td>
<td style="text-align:left;">
Update figures
</td>
</tr>
<tr>
<td style="text-align:left;">
html
</td>
<td style="text-align:left;">
<a href="https://cdn.rawgit.com/stephenslab/gtexresults/80f285faf2eaf3c1e1b4fc25317c58996d5b2525/docs/Fig.ExpressionAnalysis.html" target="_blank">80f285f</a>
</td>
<td style="text-align:left;">
Gao Wang
</td>
<td style="text-align:left;">
2017-09-20
</td>
<td style="text-align:left;">
Update figures
</td>
</tr>
</tbody>
</table>
</ul>
<p></details></p>
<hr />
<p>In this analysis, we assess whether tissue-specific eQTLs we identified can be explained by tissue-specific expression. Specifically, we take genes with tissue-specific eQTLs, and examine the distribution of expression in the eQTL-affected tissue relative to expression in other tissues.</p>
<div id="load-data-and-mash-results" class="section level2">
<h2>Load data and MASH results</h2>
<p>In the next code chunk, we load some GTEx summary statistics (average gene expression values and <em>z</em>-scores), as well as some of the results generated from the MASH analysis of the GTEx data.</p>
<p>Expression is here defined as median across individuals of the log Reads per Kilobase Mapped (RPKM).</p>
<pre class="r"><code>data &lt;- read.csv(&quot;../data/AvgExpr.csv.gz&quot;,header = TRUE)
out  &lt;- readRDS(&quot;../data/MatrixEQTLSumStats.Portable.Z.rds&quot;)
maxz &lt;- out$test.z
qtl.names &lt;-
  sapply(1:length(rownames(maxz)),
         function(x) unlist(strsplit(rownames(maxz)[x], &quot;[_]&quot;))[[1]])
rownames(data) &lt;- data[,1]
expr.data      &lt;- data[,-1]
out    &lt;- readRDS(paste(&quot;../output/MatrixEQTLSumStats.Portable.Z.coved.K3.P3&quot;,
                        &quot;lite.single.expanded.V1.posterior.rds&quot;,sep = &quot;.&quot;))
pmash &lt;- out$posterior.means
lfsr  &lt;- out$lfsr
colnames(lfsr) &lt;- colnames(pmash) &lt;- colnames(maxz)
rownames(lfsr) &lt;- rownames(pmash) &lt;- rownames(maxz)
expr.sort &lt;- expr.data[rownames(expr.data)%in%qtl.names,]
a         &lt;- match(qtl.names,rownames(expr.sort))
expr.sort &lt;- expr.sort[a,]
missing.tissues    &lt;- c(7,8,19,20,24,25,31,34,37)
exp.sort           &lt;- expr.sort[,-missing.tissues]
colnames(exp.sort) &lt;- colnames(maxz)</code></pre>
</div>
<div id="define-plotting-functions" class="section level2">
<h2>Define plotting functions</h2>
<p>Here we define a couple functions used to compare the densities and CDFs of gene expression levels.</p>
<pre class="r"><code>plot_tissuespecifictwo = function(tissuename,lfsr,ymax,curvedata,title,
                                  thresh=0.05,subset=1:44,exclude=0.01) {
  index_tissue=which(colnames(lfsr) %in% tissuename);
  ybar=title
  
  # Create a matrix showing whether or not lfsr satisfies threshold.
  sigmat = lfsr &lt;= thresh;
  sigs=which(rowSums(sigmat[,index_tissue,drop=FALSE])==length(tissuename) &amp;
             rowSums(sigmat[,-index_tissue,drop=FALSE])==0)
  norm.spec=density(curvedata[sigs,index_tissue])
  norm.other=density(curvedata[-sigs,index_tissue])
  xmin=min(norm.spec$x,norm.other$x)-1
  ymin=min(norm.spec$y,norm.other$y)
  xmax=max(norm.spec$x,norm.other$x)+1

  plot(norm.other,xlim = c(xmin,xmax),ylim=c(0,ymax),
       col=&quot;black&quot;,main=tissuename,xlab=&quot;log(RPKM)&quot;)
  
  lines(norm.spec,col=&quot;red&quot;)
  legend(&quot;right&quot;,legend = c(&quot;All Genes&quot;,&quot;Tissue Specific&quot;),
         col=c(&quot;black&quot;,&quot;red&quot;),pch=c(1,1))
  }

plot_tissuespecificthree = function(tissuename,lfsr,ymax,curvedata,title,
                                    thresh=0.05,subset=1:44,exclude=0.01) {
  index_tissue=which(colnames(lfsr) %in% tissuename);
  ybar=title
  
  # Create a matrix showing whether or not lfsr satisfies threshold.
  sigmat = lfsr &lt;= thresh
  sigs=which(rowSums(sigmat[,index_tissue,drop=FALSE])==length(tissuename) &amp;
             rowSums(sigmat[,-index_tissue,drop=FALSE])==0)
  norm.spec=ecdf(curvedata[sigs,index_tissue])
  norm.other=ecdf(curvedata[-sigs,index_tissue])
  plot(norm.other,ylim=c(0,ymax),
       col=&quot;black&quot;,main=tissuename,xlab=&quot;log(RPKM)&quot;)
  
    lines(norm.spec,col=&quot;red&quot;,cex=0.1)
    legend(&quot;right&quot;,legend = c(&quot;All Genes&quot;,&quot;Tissue Specific&quot;),
           col=c(&quot;black&quot;,&quot;red&quot;),pch=c(1,1))
  } </code></pre>
</div>
<div id="distribution-of-expression-levels-in-testis" class="section level2">
<h2>Distribution of expression levels in testis</h2>
<p>The two plots below compare the densities and cumulative distribution functions of the expression levels for all genes (black), and for genes identified as having a “tissue-specific” eQTL (red) in testis.</p>
<pre class="r"><code>plot_tissuespecifictwo(tissuename = &quot;Testis&quot;,lfsr = lfsr,
                       curvedata = log(exp.sort),title = &quot;Test&quot;,
                       thresh = 0.05 ,ymax=0.5)</code></pre>
<p><img src="figure/Fig.ExpressionAnalysis.Rmd/testis-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot_tissuespecificthree(tissuename = &quot;Testis&quot;,lfsr = lfsr,
                         curvedata = log(exp.sort),title = &quot;Test&quot;,
                         thresh = 0.05 ,ymax=1)</code></pre>
<p><img src="figure/Fig.ExpressionAnalysis.Rmd/testis-2.png" width="768" style="display: block; margin: auto;" /></p>
<p>The distribution functions are reasonably similar.</p>
</div>
<div id="distribution-of-expression-levels-in-thyroid" class="section level2">
<h2>Distribution of expression levels in thyroid</h2>
<p>Next we show the same two plots for thyroid.</p>
<pre class="r"><code>plot_tissuespecifictwo(tissuename = &quot;Thyroid&quot;,lfsr = lfsr,
                       curvedata = log(exp.sort),title = &quot;Test&quot;,
                       thresh = 0.05,ymax = 0.5)</code></pre>
<p><img src="figure/Fig.ExpressionAnalysis.Rmd/thyroid-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot_tissuespecificthree(tissuename = &quot;Thyroid&quot;,lfsr = lfsr,
                         curvedata = log(exp.sort),title = &quot;Test&quot;,
                         thresh = 0.05,ymax = 1)</code></pre>
<p><img src="figure/Fig.ExpressionAnalysis.Rmd/thyroid-2.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="distribution-of-expression-levels-in-whole-blood" class="section level2">
<h2>Distribution of expression levels in whole blood</h2>
<p>Next, we look at the distributions in whole blood cells.</p>
<pre class="r"><code>plot_tissuespecifictwo(tissuename = &quot;Whole_Blood&quot;,lfsr = lfsr,
                       curvedata = log(exp.sort),title = &quot;Test&quot;,
                       thresh = 0.05,ymax=0.5)</code></pre>
<p><img src="figure/Fig.ExpressionAnalysis.Rmd/whole-blood-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot_tissuespecificthree(tissuename = &quot;Whole_Blood&quot;,lfsr = lfsr,
                         curvedata = log(exp.sort),title = &quot;Test&quot;,
                         thresh = 0.05 ,ymax=1)</code></pre>
<p><img src="figure/Fig.ExpressionAnalysis.Rmd/whole-blood-2.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="distribution-of-expression-levels-in-fibroblasts" class="section level2">
<h2>Distribution of expression levels in fibroblasts</h2>
<p>Finally, we examine the gene expression distributions in fibroblasts.</p>
<pre class="r"><code>plot_tissuespecifictwo(tissuename = &quot;Cells_Transformed_fibroblasts&quot;,
                       lfsr = lfsr,curvedata = log(exp.sort),title = &quot;Test&quot;,
                       thresh = 0.05,ymax=0.5)</code></pre>
<p><img src="figure/Fig.ExpressionAnalysis.Rmd/fibroblasts-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot_tissuespecificthree(tissuename = &quot;Cells_Transformed_fibroblasts&quot;,
                        lfsr = lfsr,curvedata = log(exp.sort),title = &quot;Test&quot;,
                        thresh = 0.05 ,ymax=1)</code></pre>
<p><img src="figure/Fig.ExpressionAnalysis.Rmd/fibroblasts-2.png" width="768" style="display: block; margin: auto;" /></p>
<p>In each case, the distribution functions are similar. This tells us that tissue-specific eQTLs are not simply reflecting tissue-specific expression.</p>
</div>
<div id="session-information" class="section level2">
<h2>Session information</h2>
<pre class="r"><code>sessionInfo()
R version 3.4.3 (2017-11-30)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS High Sierra 10.13.4

Matrix products: default
BLAS: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
 [1] workflowr_1.0.1.9000 Rcpp_0.12.16         digest_0.6.15       
 [4] rprojroot_1.3-2      R.methodsS3_1.7.1    backports_1.1.2     
 [7] git2r_0.21.0         magrittr_1.5         evaluate_0.10.1     
[10] stringi_1.1.7        whisker_0.3-2        R.oo_1.21.0         
[13] R.utils_2.6.0        rmarkdown_1.9        tools_3.4.3         
[16] stringr_1.3.0        yaml_2.1.18          compiler_3.4.3      
[19] htmltools_0.3.6      knitr_1.20          </code></pre>
</div>

<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>
<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>

<hr>
<p>
  This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a>
  analysis was created with
  <a href="https://github.com/jdblischak/workflowr">workflowr</a> 1.0.1.9000
</p>
<hr>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
