<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Sarah Urbut, Gao Wang, Peter Carbonetto and Matthew Stephens" />


<title>Code resources accompanying Urbut *et al*, 2017</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/readable.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">MASH analysis of GTEx</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Overview</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/stephenslab/gtexresults">source</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Code resources accompanying Urbut <em>et al</em>, 2017</h1>
<h4 class="author"><em>Sarah Urbut, Gao Wang, Peter Carbonetto and Matthew Stephens</em></h4>

</div>


<p><strong>Last updated:</strong> 2018-06-01</p>
<strong>workflowr checks:</strong> <small>(Click a bullet for more information)</small>
<ul>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>R Markdown file:</strong> up-to-date </summary></p>
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Repository version:</strong> <a href="https://github.com/stephenslab/gtexresults/tree/fbc54f180666afa8a8abe18aa0388fa6651e34a0" target="_blank">fbc54f1</a> </summary></p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
<pre><code>
Ignored files:
    Ignored:    .sos/
    Ignored:    data/.sos/
    Ignored:    output/MatrixEQTLSumStats.Portable.Z.coved.K3.P3.lite.single.expanded.V1.loglik.rds
    Ignored:    workflows/.ipynb_checkpoints/
    Ignored:    workflows/.sos/

Untracked files:
    Untracked:  analysis/fastqtl2mash.Rmd
    Untracked:  gtex6_workflow_output/
    Untracked:  output/README

Unstaged changes:
    Modified:   analysis/gtex.Rmd

</code></pre>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes. </details>
</li>
</ul>
<details> <summary> <small><strong>Expand here to see past versions:</strong></small> </summary>
<ul>
<table style="border-collapse:separate; border-spacing:5px;">
<thead>
<tr>
<th style="text-align:left;">
File
</th>
<th style="text-align:left;">
Version
</th>
<th style="text-align:left;">
Author
</th>
<th style="text-align:left;">
Date
</th>
<th style="text-align:left;">
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/stephenslab/gtexresults/blob/fbc54f180666afa8a8abe18aa0388fa6651e34a0/analysis/index.Rmd" target="_blank">fbc54f1</a>
</td>
<td style="text-align:left;">
Peter Carbonetto
</td>
<td style="text-align:left;">
2018-06-01
</td>
<td style="text-align:left;">
wflow_publish(“index.Rmd”)
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/stephenslab/gtexresults/blob/2945c3d6e778179defa39527270e60ba5d21522c/analysis/index.Rmd" target="_blank">2945c3d</a>
</td>
<td style="text-align:left;">
Peter Carbonetto
</td>
<td style="text-align:left;">
2018-05-31
</td>
<td style="text-align:left;">
Removed most of the materials from the README.
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/stephenslab/gtexresults/blob/bd60ada6d8a63a28e89977bae07d6cb2e20140ea/analysis/index.Rmd" target="_blank">bd60ada</a>
</td>
<td style="text-align:left;">
Gao Wang
</td>
<td style="text-align:left;">
2018-05-31
</td>
<td style="text-align:left;">
Change workflow output folder
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/stephenslab/gtexresults/blob/380046f4d48e7a37551e9b11d7bd3ae55bde7882/analysis/index.Rmd" target="_blank">380046f</a>
</td>
<td style="text-align:left;">
Peter Carbonetto
</td>
<td style="text-align:left;">
2018-05-22
</td>
<td style="text-align:left;">
Working on improvements to Fig.GTExExamples.Rmd.
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/stephenslab/gtexresults/blob/d9994cf73798e865bd313767c5e53c1b2dde9c63/analysis/index.Rmd" target="_blank">d9994cf</a>
</td>
<td style="text-align:left;">
Peter Carbonetto
</td>
<td style="text-align:left;">
2018-05-16
</td>
<td style="text-align:left;">
A few very minor misc. changes.
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/stephenslab/gtexresults/blob/3431f022aad82593e812d1e82df41d7b972a591c/analysis/index.Rmd" target="_blank">3431f02</a>
</td>
<td style="text-align:left;">
Peter Carbonetto
</td>
<td style="text-align:left;">
2018-05-14
</td>
<td style="text-align:left;">
wflow_publish(“index.Rmd”)
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/stephenslab/gtexresults/blob/ea42e8a34505bf6d32b37485659b2a9bdbb6ccf8/analysis/index.Rmd" target="_blank">ea42e8a</a>
</td>
<td style="text-align:left;">
Peter Carbonetto
</td>
<td style="text-align:left;">
2018-05-07
</td>
<td style="text-align:left;">
Added step to install packages in instructions (index.Rmd).
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/stephenslab/gtexresults/blob/6ebb3f99e98829b3d132656e0856920be9523147/analysis/index.Rmd" target="_blank">6ebb3f9</a>
</td>
<td style="text-align:left;">
Peter Carbonetto
</td>
<td style="text-align:left;">
2018-05-07
</td>
<td style="text-align:left;">
wflow_publish(“index.Rmd”)
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/stephenslab/gtexresults/blob/c6c84508b4028a30e3db43a8afcffaa37c57be29/analysis/index.Rmd" target="_blank">c6c8450</a>
</td>
<td style="text-align:left;">
Peter Carbonetto
</td>
<td style="text-align:left;">
2018-05-07
</td>
<td style="text-align:left;">
Setting up workflowr settings.
</td>
</tr>
<tr>
<td style="text-align:left;">
html
</td>
<td style="text-align:left;">
<a href="https://cdn.rawgit.com/stephenslab/gtexresults/c6c84508b4028a30e3db43a8afcffaa37c57be29/docs/index.html" target="_blank">c6c8450</a>
</td>
<td style="text-align:left;">
Peter Carbonetto
</td>
<td style="text-align:left;">
2018-05-07
</td>
<td style="text-align:left;">
Setting up workflowr settings.
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/stephenslab/gtexresults/blob/33f72d6007c845a7f0e5b1610ec4d9b721cca299/analysis/index.Rmd" target="_blank">33f72d6</a>
</td>
<td style="text-align:left;">
Peter Carbonetto
</td>
<td style="text-align:left;">
2018-05-07
</td>
<td style="text-align:left;">
Added workflowr 1.0.1.9000 files.
</td>
</tr>
<tr>
<td style="text-align:left;">
html
</td>
<td style="text-align:left;">
<a href="https://cdn.rawgit.com/stephenslab/gtexresults/b4e38c281089fb599eac5b70930ff8366b1a6f48/docs/index.html" target="_blank">b4e38c2</a>
</td>
<td style="text-align:left;">
Gao Wang
</td>
<td style="text-align:left;">
2017-09-20
</td>
<td style="text-align:left;">
Fix one more broken link
</td>
</tr>
<tr>
<td style="text-align:left;">
html
</td>
<td style="text-align:left;">
<a href="https://cdn.rawgit.com/stephenslab/gtexresults/80f285faf2eaf3c1e1b4fc25317c58996d5b2525/docs/index.html" target="_blank">80f285f</a>
</td>
<td style="text-align:left;">
Gao Wang
</td>
<td style="text-align:left;">
2017-09-20
</td>
<td style="text-align:left;">
Update figures
</td>
</tr>
<tr>
<td style="text-align:left;">
html
</td>
<td style="text-align:left;">
<a href="https://cdn.rawgit.com/stephenslab/gtexresults/afc401f3e92f562d878d377ddac5033f230e7c51/docs/index.html" target="_blank">afc401f</a>
</td>
<td style="text-align:left;">
Peter Carbonetto
</td>
<td style="text-align:left;">
2017-09-20
</td>
<td style="text-align:left;">
Moved doc to docs.
</td>
</tr>
</tbody>
</table>
</ul>
<p></details></p>
<hr />
<div id="overview" class="section level2">
<h2>Overview</h2>
<p>This repository contains code resources to accompany our research paper:</p>
<blockquote>
<p>Sarah M. Urbut, Gao Wang and Matthew Stephens (2017). <em>Flexible statistical methods for estimating and testing effects in genomic studies with multiple conditions.</em> bioRxiv <a href="http://dx.doi.org/10.1101/096552">doi:10.1101/096552</a>.</p>
</blockquote>
<p>We provide three sets of resources:</p>
<ol style="list-style-type: decimal">
<li><p>If you are primarily interested in applying multivariate adaptive shrinkage (<em>mash</em>) methods to your own data, please see the <a href="https://github.com/stephenslab/mashr">mashr repository</a>.</p></li>
<li><p>If you are working with association statistics from the <a href="http://gtexportal.org">GTEx study</a>, or a similar data set, please see here for our code to convert association statistics in the <a href="http://fastqtl.sourceforge.net">FastQTL</a> format, or a similar format, to a format that is more convenient for analysis with mashr.</p></li>
<li><p>If you would like follow the exact steps we took to produce the results presented in the manuscript, follow the instructions <a href="gtex.html">here</a>.</p></li>
</ol>
</div>
<div id="citing-this-work" class="section level2">
<h2>Citing this work</h2>
<p>If you find any of the source code in this repository useful for your work, please cite our paper:</p>
<blockquote>
<p>Sarah M. Urbut, Gao Wang and Matthew Stephens (2017). <em>Flexible statistical methods for estimating and testing effects in genomic studies with multiple conditions.</em> bioRxiv <a href="http://dx.doi.org/10.1101/096552">doi:10.1101/096552</a>.</p>
</blockquote>
</div>
<div id="license" class="section level2">
<h2>License</h2>
<p>Copyright (c) 2016-2018, Sarah Urbut, Gao Wang, Peter Carbonetto and Matthew Stephens.</p>
<p>All source code and software in this repository are made available under the terms of the <a href="https://opensource.org/licenses/mit-license.html">MIT license</a>. See the <a href="LICENSE" class="uri">LICENSE</a> file for the full text of the license.</p>
</div>
<div id="contributors" class="section level2">
<h2>Contributors</h2>
<p>This project was developed by <a href="https://github.com/surbut">Sarah Urbut</a>, <a href="https://github.com/gaow">Gao Wang</a>, <a href="http://pcarbo.github.io">Peter Carbonetto</a>. <a href="https://stephenslab.uchicago.edu">Matthew Stephens</a> at the University of Chicago, with contributions from</p>
<p>We provide three sets of code resources to accompany Urbut, Wang, Carbonetto &amp; Stephens (2017)</p>
<p>This site contains source code to produce results from Urbut, Wang &amp; Stephens (2017).</p>
<p>If you are primarily interested in applying the statistical methods to your own data, please see the <a href="https://github.com/stephenslab/mashr">mashr package</a>.</p>
<p>To reproduce the results of Urbut, Wang &amp; Stephens (2017), please follow the instructions <a href="https://github.com/stephenslab/gtexresults">here</a>.</p>
</div>
<div id="overview-1" class="section level2">
<h2>Overview</h2>
<p>To reproduce the results of Urbut, Wang &amp; Stephens (2017), please follow these instructions.</p>
<p>For more information, please see the <a href="https://github.com/stephenslab/gtexresults">README</a>.</p>
<p>The complete analyses require installation of several programs and libraries, and requires several large data sets. To facilitate reproducing our results, we provide pre-processed data for use with the core analysis, and a bioinformatics pipeline with a small toy data-set to demonstrate the pre-processing step. We have also developed a <a href="https://hub.docker.com/r/gaow/mash-paper">Docker container</a> that includes all software components necessary to run the analyses. Docker can run on most popular operating systems (Mac, Windows and Linux). It also runs on cloud computing services such as Amazon Web Services and Microsoft Azure. If you have not used Docker before, you might want to read <a href="https://docs.docker.com/engine/docker-overview">this</a> to learn the basic concepts and understand the benefits of Docker.</p>
<p>For details on how the Docker image was configured, see the <a href="workflows/Dockerfile">Dockerfile</a>. The Docker image used for our analyses is based on <a href="https://hub.docker.com/r/gaow/lab-base">gaow/lab-base</a>, a Docker image for development with R and Python.</p>
<p>If you prefer to run the analyses without Docker, <em>add a few details about where you can find out more about software and libraries used, and other computing environment setup steps (mention Python 3.x, R, SFA, ExtremeDeconvolution, MOSEK, OpenMP, MKL, GSL, HDF5 tools, pytables rhdf5, and for an improved MASH implementation mashr is also installed).</em></p>
</div>
<div id="download-and-install-docker" class="section level2">
<h2>1. Download and install Docker</h2>
<p>Download <a href="https://docs.docker.com/install">Docker</a> (note that a free <a href="https://www.docker.com/community-edition">community edition</a> of Docker is available), and install it following the instructions provided on the Docker website. Once you have installed Docker, check that Docker is working correctly by following <a href="https://docs.docker.com/get-started">Part 1 of the Getting Started guide</a>. If you are using Docker for the first time, we recommend reading the entire Getting Started guide. <em>Note that setting up Docker requires that you have administrator access to your computer.</em> (<a href="https://singularity.lbl.gov/docs-docker">Singularity</a> is an alternative that accepts Docker images and does not require administrator access.)</p>
</div>
<div id="download-and-test-docker-image" class="section level2">
<h2>2. Download and test Docker image</h2>
<p>Run this <code>alias</code> command in the shell, which will be used below to run commands inside the Docker container:</p>
<pre class="bash"><code>alias mash-docker=&#39;docker run --security-opt label:disable -t -P -h MASH &#39;\
&#39;-w $PWD -v $HOME:/home/$USER -v /tmp:/tmp -v $PWD:$PWD &#39;\
&#39;-u $UID:${GROUPS[0]} -e HOME=/home/$USER -e USER=$USER gaow/mash-paper&#39;</code></pre>
<p>The <code>-v</code> flags in this command map directories between the standard computing environment and the Docker container. Since the analyses below will write files to these directories, it is important to ensure that:</p>
<ul>
<li><p>Environment variables <code>$HOME</code> and <code>$PWD</code> are set to valid and writeable directories (usually your home and current working directories, respectively).</p></li>
<li><p><code>/tmp</code> should also be a valid and writeable directory.</p></li>
</ul>
<p>If any of these statements are not true, please adjust the <code>alias</code> accordingly. The remaining options only affect operation of the container, and so should function the same regardless of your operating system.</p>
<p>Next, run a simple command in the Docker container to check that has loaded successfully:</p>
<pre><code>mash-docker uname -sn</code></pre>
<p>This command will download the Docker image if it has not already been downloaded.</p>
<p>If the container was successfully run, you should see this information about the Docker container outputted to the screen:</p>
<pre><code>Linux MASH</code></pre>
<p>You can also run these commands to show the information about the image downloaded to your computer and the container that has run (and exited):</p>
<pre class="bash"><code>docker image list
docker container list --all</code></pre>
<p><em>Note:</em> If you get error “Cannot connect to the Docker daemon. Is the docker daemon running on this host?” in Linux or macOS, see <a href="https://askubuntu.com/questions/477551/how-can-i-use-docker-without-sudo">here for Linux</a> or <a href="https://github.com/wodby/docker4drupal/issues/15">here for Mac</a> for suggestions on how to resolve this issue.</p>
</div>
<div id="clone-or-download-this-repository" class="section level2">
<h2>3. Clone or download this repository</h2>
<p>Clone or download the <code>gtexresults</code> repository to your computer, then change your working directory in the shell to the root of the repository, e.g.,</p>
<pre class="bash"><code>cd gtexresults</code></pre>
<p>After doing this, running <code>ls -1</code> should show the top-level contents of this repository:</p>
<pre><code>LICENSE
README.md
TODO.txt
analysis
data
docs
output
workflows</code></pre>
<p>All commands below will be run from this directory.</p>
<div id="fit-mash-model-and-compute-posterior-statistics" class="section level3">
<h3>4. Fit MASH model and compute posterior statistics</h3>
<p>Assuming your working directory is the root of the git repository (you can check by running <code>pwd</code>), run all the steps of the analysis with this command:</p>
<pre class="bash"><code>mash-docker sos run workflows/gtex6_mash_analysis.ipynb</code></pre>
<p>This command will take several hours to run—see below for more information on the individual steps. All outputs generated by this command will be saved to folder <code>output</code> inside the repository.</p>
<p>Note that you may recognize this file as a Jupyter notebook. Indeed, you may open this notebook in Jupyter. However, you should not step through the code sequentially as you would in a typical Jupyter notebook; this is because the code in this notebook is meant to be run using the <a href="https://github.com/vatlab/SoS">Script of Scripts (SoS)</a> framework.</p>
<p>This command will execute the following steps of the analysis:</p>
<ul>
<li><p>Compute a sparse factorization of the (centered) z-scores using the <a href="http://stephenslab.uchicago.edu/software.html#sfa">SFA software</a>, with K = 5 factors, and save the factors in an <code>.rds</code> file. This will be used to construct the mixture-of-multivariate normals prior. This step is labeled <code>sfa</code>, and should only take a few minutes to run.</p></li>
<li><p>Compute additional “data-driven” prior matrices by computing a singular value decomposition of the (centered) z-scores and low-rank approximations to the empirical covariance matrices. Most of the work in this step involves running the Extreme Deconvolution method. The outcome of running the Extreme Deconvolution method is saved to a new <code>.rds</code> file. This step is labeled <code>mash-paper_1</code> and may take several hours to run (in one run on a MacBook Pro with a 3.5 GHz Intel Core i7, it took over 6 hours to complete).</p></li>
<li><p>A final collection of “canonical” and single-rank prior matrices based on SFA and the “BMAlite” models of Flutre <em>et al</em> (2013). These matrices are again written to another <code>.rds</code> file. This step is labeled <code>mash-paper_2</code>, and should take at most a minute to run.</p></li>
<li><p>The <code>mash-paper_3</code> step fits the MASH (“multivariate adaptive shrinkage”) model to the GTEx data (the centered z-scores); the model parameters estimated in this fitting step are the weights of the multivariate normal mixture. The outputs from this step are the estimated mixture weights and the conditional likelihood matrix. These two outputs are saved to two separate <code>.rds</code> files. This step is expected to take at most a few hours to complete.</p></li>
<li><p>The <code>mash-paper_4</code> step computes posterior statistics using the fitted MASH model from the previous step. These posterior statistics are summarized and visualized in subsequent analyses. The posterior statistics are saved to another <code>.rds</code> file. This step is expected to take a few hours to complete.</p></li>
</ul>
<p>Finally, note that all containers that have run and exited will still be retained in the Docker system. Run <code>docker container list --all</code> to list all previous run containers. To clear these previously run containers, run <code>docker container prune</code>. See <a href="https://stackoverflow.com/questions/17014263/should-i-be-concerned-about-excess-non-running-docker-containers">here</a> for more information.</p>
</div>
<div id="add-step-5-title-here" class="section level3">
<h3>5. Add Step 5 title here</h3>
<p>Install some packages from CRAN:</p>
<pre class="r"><code># Add commands here to install packages.</code></pre>
<p><strong>For convenience, the results needed to generate the figures and tables have been saved in the <code>output</code> folder.</strong></p>
<p><strong>FIXME: update figure plotting instructions</strong></p>
<p>The input data necessary to run this analysis is all available under inputs. This may take some time to run. We have provided the outputs of running mash in <code>Data_vhat</code>.</p>
<p>This repo is organized so that you can run Mash using the gteX data contained in <strong>Inputs</strong> to produce the parameters and posteriors from mashr.</p>
<p>The directory <strong>Plots_for_Paper_vmat</strong> contains .Rmd files to plot figures from the paper, using our results which are provided in <strong>Results_Data</strong>.</p>
<p>Figure 3:<a href="Fig.Uk3.html">Summary of primary patterns identified by mash in GTEx data</a></p>
<p>Figure 4:<a href="Fig.GTExExamples.html">Examples illustrating of how mash uses patterns of sharing to inform effect estimates in the GTEx data.</a></p>
<p>Figure 5:<a href="Fig.SharingHist.pdf">Histogram of Sharing</a></p>
<p>Figure 6:<a href="Fig.SharingMag.html">Pairwise sharing by magnitude of eQTL among tissues</a></p>
<p>Supplementary Figure 1:<a href="Fig.SampleSize.html">Sample sizes and effective sample sizes from mash analysis across tissues</a></p>
<p>Supplementary Figure 2:There are 4 figures here:</p>
<p><a href="Fig.Uk2.pdf">Summary of covariance matrices Uk with largest estimated weight (&gt; 1%) in GTEx data:Uk2</a></p>
<p><a href="Fig.Uk4.html">Summary of covariance matrices Uk with largest estimated weight (&gt; 1%) in GTEx data:Uk4</a></p>
<p><a href="Fig.Uk5.html">Summary of covariance matrices Uk with largest estimated weight (&gt; 1%) in GTEx data:Uk5</a></p>
<p><a href="Fig.Uk8.html">Summary of covariance matrices Uk with largest estimated weight (&gt; 1%) in GTEx data:Uk8</a></p>
<p>Supplementary Figure 3: Illustration of how Linkage Disequilibrium can impact effect estimate <a href="TwoSNP.html">table</a> and <a href="TwoSNPPlot.pdf">figure</a></p>
<p>Supplementary Figure 4:<a href="Fig.SharingSign.html">Pairwise Sharing By Sign</a></p>
<p>Supplementary Figure 5:<a href="Fig.Tspecific.html">Number of “tissue-specific eQTLs” in each tissues.</a></p>
<p>Supplementary Figure 6:<a href="Fig.ExpressionAnalysis.html">Expression levels in genes with “tissue-specific eQTLs” are similar to those in other genes</a></p>
<p>Table 1: Heterogeneity Analysis <a href="Table.hettablesim.html">Simulation</a> and <a href="Table.HeterogeneityTables.html">Data</a>.</p>
</div>
</div>
<div id="more-detailed-usage-notes" class="section level2">
<h2>More detailed usage notes</h2>
<p>Above we have given the minimal instructions necessary to reproduce the results of Urbut <em>et al</em> (2017). Here are some additional details about the analyses.</p>
<p><em>TO DO: Things that will go here:</em></p>
<ul>
<li><p>Explain how to get a summary of the possible analysis steps that can be run.</p></li>
<li><p>See the Jupyter notebook to get more details; how the notebook should be interpreted.</p></li>
<li><p>Explain how to run the analysis using the improved (faster) implementation of the <a href="https://github.com/stephenslab/mashr">mashr R package</a>.</p></li>
<li><p><em>MOSEK?</em></p></li>
<li><p>In the <a href="data" class="uri">data</a> folder, we have provided a file <a href="data/MatrixEQTLSumStats.Portable.Z.rds">MatrixEQTLSumStats.Portable.Z.rds</a> containing eQTL summary statistics in a format convenient for running MASH. This was generated from the original eQTL summary statistics downloaded from the GTEx Portal website, then converted using the code in <code>fastqtl_to_mash.ipynb</code>. See below for details on this step.</p></li>
</ul>
</div>
<div id="converting-eqtl-summary-statistics-to-mash-format" class="section level2">
<h2>Converting eQTL summary statistics to MASH format</h2>
<p>Here we explain how the <a href="data/MatrixEQTLSumStats.Portable.Z.rds">MatrixEQTLSumStats.Portable.Z.rds</a> data file was generated from the source data downloaded from the <a href="http://gtexportal.org">GTEx Portal</a>. The source data are the SNP-gene association statistics from release 6 of the GTEx Project (<code>GTEx_Analysis_V6_all-snp-gene-associations.tar</code>).</p>
<p>Under the repo you will find <code>workflows/fastqtl_to_mash.ipynb</code> to convert eQTL summary statistics (default to <code>fastqtl</code> format) to MASH format. Computation is configured to run in parallel for eQTL results from multiple studies. Example data-set can be found at <code>data/eQTLDataDemo</code>. The workflow file is documented in itself, and has a few options to customize the input and output.</p>
<p>To read what’s available, run:</p>
<pre class="bash"><code>mash-docker sos run workflows/fastqtl_to_mash.ipynb export</code></pre>
<p>and read the HTML files <code>output/fastqtl_to_mash.lite.html</code> and <code>output/fastqtl_to_mash.full.html</code></p>
<p>To run the conversion:</p>
<pre class="bash"><code>mash-docker sos run workflows/fastqtl_to_mash.ipynb \
  --data_list data/eQTLDataDemo/FastQTLSumStats.list \
  --gene_list data/eQTLDataDemo/GTEx_genes.txt</code></pre>
<p>In practice for GTEx data the conversion is computationally intensive and is best done on a cluster environment with <a href="https://vatlab.github.io/sos-docs/doc/documentation/Remote_Execution.html">configurations to run the workflow across different nodes</a>.</p>
</div>
<div id="developer-notes" class="section level2">
<h2>Developer notes</h2>
<p>Run the following command to update the Docker image:</p>
<pre class="bash"><code>docker pull gaow/mash-paper</code></pre>
</div>

<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>
<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>

<hr>
<p>
  This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a>
  analysis was created with
  <a href="https://github.com/jdblischak/workflowr">workflowr</a> 1.0.1.9000
</p>
<hr>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
